src/PointwiseFunctions/AnalyticData/GrMhd/MagneticRotor.cpp

```diff
@@ -0,0 +1,211 @@
+
+#include <cmath>  // IWYU pragma: keep
+#include <pup.h>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "Parallel/PupStlCpp11.hpp"
+#include "PointwiseFunctions/Hydro/LorentzFactor.hpp"
+#include "Utilities/ConstantExpressions.hpp"  // IWYU pragma: keep
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/Math.hpp"  // IWYU pragma: keep
+
+// IWYU pragma: no_include <complex>
+
+namespace {
+template <typename DataType>
+Scalar<DataType> compute_piecewise(
+    const tnsr::I<DataType, 3, Frame::Inertial>& x, const double rotor_radius,
+    const double rotor_value, const double background_value) noexcept {
+  const DataType cylindrical_radius =
+      sqrt(square(get<0>(x)) + square(get<1>(x)));
+  return Scalar<DataType>(rotor_value -
+                          (rotor_value - background_value) *
+                              step_function(cylindrical_radius - rotor_radius));
+}
+}  // namespace
+
+/// \cond
+namespace grmhd {
+namespace AnalyticData {
+
+MagneticRotor::MagneticRotor(const double rotor_radius,
+                             const double rotor_density,
+                             const double background_density,
+                             const double pressure,
+                             const double angular_velocity,
+                             const std::array<double, 3> magnetic_field,
+                             const double adiabatic_index)
+    : rotor_radius_(rotor_radius),
+      rotor_density_(rotor_density),
+      background_density_(background_density),
+      pressure_(pressure),
+      angular_velocity_(angular_velocity),
+      magnetic_field_(magnetic_field),
+      adiabatic_index_(adiabatic_index),
+      equation_of_state_(adiabatic_index) {}
+
+void MagneticRotor::pup(PUP::er& p) noexcept {
+  p | rotor_radius_;
+  p | rotor_density_;
+  p | background_density_;
+  p | pressure_;
+  p | angular_velocity_;
+  p | magnetic_field_;
+  p | adiabatic_index_;
+  p | equation_of_state_;
+  p | background_spacetime_;
+}
+
+template <typename DataType>
+tuples::TaggedTuple<hydro::Tags::RestMassDensity<DataType>>
+MagneticRotor::variables(
+    const tnsr::I<DataType, 3, Frame::Inertial>& x,
+    tmpl::list<hydro::Tags::RestMassDensity<DataType>> /*meta*/) const
+    noexcept {
+  return compute_piecewise(x, rotor_radius_, rotor_density_,
+                           background_density_);
+}
+
+template <typename DataType>
+tuples::TaggedTuple<hydro::Tags::SpatialVelocity<DataType, 3>>
+MagneticRotor::variables(
+    const tnsr::I<DataType, 3, Frame::Inertial>& x,
+    tmpl::list<hydro::Tags::SpatialVelocity<DataType, 3>> /*meta*/) const
+    noexcept {
+  Scalar<DataType> angular_velocity =
+      compute_piecewise(x, rotor_radius_, angular_velocity_, 0.0);
```

Comment:
I would add a check that prevents `rotor_radius_ * angular_velocity_` being greater than one